Replace mutable github action tags with commit SHAs

levi178u opened the pull request:
**Purpose of PR?**:

parts #2736 

Changes done:
- Replaces all mutable `@v*` tags in GitHub Actions workflows with exact commit SHAs .
- Referencing tags like `actions/checkout@v4`) is vulnerable to supply chain attacks, as tags are mutable.

levi178u committed:
Replace mutable github action tags with commit SHAs

Replaces mutable github action tags with immutable commit SHAs for preventing third party interference.

levi178u commented:
@matrss Pls review this PR .

ReimarBauer commented on .github/workflows/build_docs_gallery.yml:
```diff
@@ -7,8 +7,8 @@ jobs:
     steps:
-    - uses: actions/checkout@v4
-    - uses: prefix-dev/setup-pixi@v0.8.4
+    - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
```
this needs to be done automated.

ReimarBauer reviewed (changes requested):
we need a way to let the bot ci do that automated

levi178u commented:
> we need a way to let the bot ci do that automated

Means we need to write code for enabling the ci bot to fetch it automatically without any user being involved.

levi178u commented:
Also how can we test it that the version tags are being updated with SHA keys by  the ci bot ?

levi178u commented:
Also @ReimarBauer can u share with me the github actions tools or resource, using which the ci bot would be able to replace `@vXYZ` with `@SHA   # vXYZ`

matrss reviewed (changes requested):
Dependabot should already be able to update these hashes based on a version tag that is indicated by a comment on the same line; see the issue for reference, dependabot should be able to work with that format. This PR is missing changes to all of the other workflows we have, it currently only touches one.

levi178u commented:
> Dependabot should already be able to update these hashes based on a version tag that is indicated by a comment on the same line; see the issue for reference, dependabot should be able to work with that format. This PR is missing changes to all of the other workflows we have, it currently only touches one.

- Yes I noticed the hashes part needs to be added for the bot to figure out the version tags , replace `@vXYZ` with `@SHA   # vXYZ`
- may I know what other changes I need to do and also can u share with me the github actions tools or resource for this action to be automated.

matrss commented:
> * may I know what other changes I need to do and also can u share with me the github actions tools or resource for this action to be automated.

There are more files in .github/workflows that need to be fixed. The only automation we need after the references are changed to hashes is the already configured dependabot.

levi178u commented:
> > * may I know what other changes I need to do and also can u share with me the github actions tools or resource for this action to be automated.
> 
> There are more files in .github/workflows that need to be fixed. The only automation we need after the references are changed to hashes is the already configured dependabot.

Got it lemme make the changes and push them

levi178u committed:
Replaced all version tag references with corresponding SHA commit hashes

 Replaced all version tag references with corresponding SHA commit hashes in .github/workflows

levi178u commented:
Mentors pls review the changes now

ReimarBauer commented:
Hash pin got integrated into dependabot, what is different in our configuration? 

https://github.com/pypa/packaging/issues/756

ReimarBauer reviewed (changes requested):
the PR needs to become rebased, the action version at least the comment is outdated

levi178u committed:
Merge branch 'stable' into 2736-immutable-sha-key-refs
